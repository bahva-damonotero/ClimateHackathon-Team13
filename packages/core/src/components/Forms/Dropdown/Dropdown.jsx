import React from 'react';
import PropTypes from 'prop-types';
import isEqual from 'lodash.isequal';

import { FormError, Label } from '..';

export default function Dropdown({
  children,
  comparator,
  error,
  id,
  onChange,
  optionRenderer,
  options,
  showSelectOption,
  value,
  ...props
}) {
  function selectionIndex() {
    const index = options.findIndex((option) => {
      if (comparator) return comparator(option, value);

      if (typeof value === 'object') return isEqual(option, value);

      return option === value;
    });

    return index >= 0 ? index : undefined;
  }

  return (
    <>
      <Label error={error} target={id}>
        {children}
      </Label>

      <FormError error={error} />

      {/* eslint-disable jsx-a11y/no-onchange */}
      <select
        className="usa-select"
        id={id}
        name={id}
        onChange={
          onChange
            ? (e) => {
                const event = new CustomEvent('change', {
                  detail: {
                    name: id,
                    value: options[e.target.value]
                  }
                });
                onChange(event);
              }
            : undefined
        }
        value={selectionIndex()}
        {...props}
      >
        {showSelectOption && <option value="">- Select -</option>}

        {options.map((option, index) => (
          <option key={`${id}--${index}`} value={index}>
            {optionRenderer ? optionRenderer(option) : option}
          </option>
        ))}
      </select>
      {/* eslint-enable jsx-a11y/no-onchange */}
    </>
  );
}

Dropdown.propTypes = {
  /** Label for input */
  children: PropTypes.node.isRequired,
  /**
   * Let's you specify a custom function for comparing the selected value with the list of options.
   * If this is not provided, lodash's isEqual function is used: https://lodash.com/docs/#isEqual.
   */
  comparator: PropTypes.func,
  /** Error message to display */
  error: PropTypes.string,
  /** ID of the input */
  id: PropTypes.string.isRequired,
  /** Function called whenever the dropdown selection is changed */
  onChange: PropTypes.func,
  /** Function that when given an option should return how the option should be displayed in the dropdown menu */
  optionRenderer: PropTypes.func,
  /** Array of dropdown options */
  options: PropTypes.array.isRequired,
  /** Should the "- Select -" option be shown? */
  showSelectOption: PropTypes.bool,
  /** Used to set the dropdown selection */
  value: PropTypes.any
};

Dropdown.defaultProps = {
  showSelectOption: true
};
