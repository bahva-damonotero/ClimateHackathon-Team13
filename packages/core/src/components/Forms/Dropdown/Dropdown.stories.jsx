import React from 'react';
import { action } from '@storybook/addon-actions';

import { Form } from '..';
import Dropdown from '.';

export default {
  title: 'Forms/Dropdown',
  component: Dropdown
};

const Template = (args) => (
  <Form>
    <Dropdown {...args} />
  </Form>
);

const defaultArgs = {
  children: 'Label',
  id: 'dropdown-example',
  options: [
    {
      firstName: 'John',
      lastName: 'Doe'
    },
    {
      firstName: 'Jane',
      lastName: 'Doe'
    },
    {
      firstName: 'Bob',
      lastName: 'Bobson'
    }
  ],
  optionRenderer: ({ firstName, lastName }) => `${firstName} ${lastName}`,
  onChange: action('onChange')
};

export const def = Template.bind({});
def.storyName = 'Default';
def.args = {
  ...defaultArgs
};

export const selection = Template.bind({});
selection.args = {
  ...defaultArgs,
  value: defaultArgs.options[1],
  onChange: undefined,
  readOnly: true
};

export const noSelectOption = Template.bind({});
noSelectOption.storyName = 'No "- Select -" option';
noSelectOption.args = {
  ...defaultArgs,
  showSelectOption: false
};

export const error = Template.bind({});
error.args = {
  ...defaultArgs,
  error: 'Invalid selection'
};

export const customComparatorFunction = Template.bind({});
customComparatorFunction.args = {
  ...defaultArgs,
  comparator: (value, option) =>
    value.firstName === option.firstName && value.lastName === option.lastName,
  value: defaultArgs.options[2],
  onChange: undefined,
  readOnly: true
};

export const stringArray = Template.bind({});
stringArray.args = {
  ...defaultArgs,
  options: ['John Doe', 'Jane Doe', 'Bob Bobson'],
  optionRenderer: undefined
};
