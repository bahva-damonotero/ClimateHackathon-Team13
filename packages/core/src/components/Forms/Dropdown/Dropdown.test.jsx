/* eslint-disable import/first */
jest.mock('lodash.isequal');

import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import isEqual from 'lodash.isequal';

import Dropdown from '.';

describe('Dropdown', () => {
  const id = 'dropdown-example';
  const label = 'Dropdown Label';

  const basicOptions = ['John Doe', 'Jane Doe', 'Bob Bobson'];

  const options = [
    { firstName: 'John', lastName: 'Doe' },
    { firstName: 'Jane', lastName: 'Doe' },
    { firstName: 'Bob', lastName: 'Bobson' }
  ];

  const optionRenderer = ({ firstName, lastName }) =>
    `${firstName} ${lastName}`;

  const comparator = (value, option) =>
    value && option
      ? value.firstName === option.firstName &&
        value.lastName === option.lastName
      : false;

  const SELECT_SELECTOR = `#${id}.usa-select`;
  const LABEL_SELECTOR = '.usa-label';

  it('should render correctly', () => {
    const { container } = render(
      <Dropdown id={id} options={basicOptions}>
        {label}
      </Dropdown>
    );

    expect(container.querySelector(SELECT_SELECTOR)).toBeTruthy();

    const labelNode = container.querySelector(LABEL_SELECTOR);
    expect(labelNode.innerHTML).toBe(label);
    expect(labelNode.getAttribute('for')).toBe(id);
  });

  it('should call `onChange` when the value changes', () => {
    const onChange = jest.fn();

    const { container } = render(
      <Dropdown
        comparator={comparator}
        id={id}
        onChange={onChange}
        optionRenderer={optionRenderer}
        options={options}
        value={options[1]}
      >
        {label}
      </Dropdown>
    );

    const selectNode = container.querySelector(SELECT_SELECTOR);

    fireEvent.change(selectNode, {
      target: {
        value: 0
      }
    });

    expect(onChange).toHaveBeenCalledWith(
      new CustomEvent('change', {
        detail: {
          name: id,
          value: options[0]
        }
      })
    );
  });

  it('should use `lodash.isEqual` when `comparator` is not provided', () => {
    // render the dropdown with no value
    const initialProps = {
      id,
      onChange: () => {},
      optionRenderer,
      options
    };

    const { rerender } = render(<Dropdown {...initialProps}>{label}</Dropdown>);

    // update the dropdown's value programmatically/externally
    const newValue = options[0];

    rerender(
      <Dropdown {...initialProps} value={newValue}>
        {label}
      </Dropdown>
    );

    expect(isEqual).toHaveBeenCalled();
  });

  it('should handle falsey values based on given options', () => {
    const initialProps = {
      id,
      onChange: () => {},
      options: ['string', 0, false]
    };

    const { container, rerender } = render(
      <Dropdown {...initialProps}>{label}</Dropdown>
    );

    expect(container.querySelector(SELECT_SELECTOR).value).toBe('');

    rerender(
      <Dropdown {...initialProps} value={0}>
        {label}
      </Dropdown>
    );

    expect(container.querySelector(SELECT_SELECTOR).value).toBe('1');

    rerender(
      <Dropdown {...initialProps} value={false}>
        {label}
      </Dropdown>
    );

    expect(container.querySelector(SELECT_SELECTOR).value).toBe('2');
  });

  it('should match snapshot', () => {
    const tree = renderer
      .create(
        <Dropdown id={id} options={basicOptions}>
          {label}
        </Dropdown>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
