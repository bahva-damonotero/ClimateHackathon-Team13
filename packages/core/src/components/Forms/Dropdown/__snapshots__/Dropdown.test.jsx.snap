// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should match snapshot 1`] = `
Array [
  <label
    className="usa-label"
    htmlFor="dropdown-example"
  >
    Dropdown Label
  </label>,
  <select
    className="usa-select"
    id="dropdown-example"
    name="dropdown-example"
  >
    <option
      value=""
    >
      - Select -
    </option>
    <option
      value={0}
    >
      John Doe
    </option>
    <option
      value={1}
    >
      Jane Doe
    </option>
    <option
      value={2}
    >
      Bob Bobson
    </option>
  </select>,
]
`;
